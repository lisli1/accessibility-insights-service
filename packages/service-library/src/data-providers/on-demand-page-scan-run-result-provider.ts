// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { inject, injectable } from 'inversify';

import { GuidUtils, HashGenerator } from 'common';

import { CosmosContainerClient, cosmosContainerClientTypes } from 'azure-services';
import { flatMap, groupBy } from 'lodash';
import { ItemType, OnDemandPageScanResult } from 'storage-documents';

@injectable()
export class OnDemandPageScanRunResultProvider {
    public static readonly partitionKeyPreFix = 'pageScanRunResult';

    constructor(
        @inject(HashGenerator) private readonly hashGenerator: HashGenerator,
        @inject(GuidUtils) private readonly guidUtils: GuidUtils,
        @inject(cosmosContainerClientTypes.OnDemandScanRunsCosmosContainerClient)
        private readonly cosmosContainerClient: CosmosContainerClient,
    ) {}

    public async createScanRuns(scanRuns: OnDemandPageScanResult[]): Promise<void> {
        const scanRunsByPartition = groupBy(scanRuns, scanRun => {
            this.setSystemProperties(scanRun);

            return scanRun.partitionKey;
        });

        await Promise.all(
            Object.keys(scanRunsByPartition).map(async pKey => {
                return this.cosmosContainerClient.writeDocuments(scanRunsByPartition[pKey], pKey);
            }),
        );
    }

    public async readScanRuns(scanIds: string[]): Promise<OnDemandPageScanResult[]> {
        const scanIdsByPartition = groupBy(scanIds, scanId => {
            return this.getPartitionKey(scanId);
        });

        const response = await Promise.all(
            Object.keys(scanIdsByPartition).map(async pKey => {
                return this.cosmosContainerClient.executeQueryWithContinuationToken<OnDemandPageScanResult>(async token => {
                    return this.cosmosContainerClient.queryDocuments<OnDemandPageScanResult>(
                        this.getReadScanQueryForScanIds(scanIdsByPartition[pKey]),
                        token,
                        pKey,
                    );
                });
            }),
        );

        return flatMap(response);
    }

    public async updateScanRun(pageScanResult: OnDemandPageScanResult): Promise<void> {
        this.setSystemProperties(pageScanResult);

        await this.cosmosContainerClient.mergeOrWriteDocument(pageScanResult);
    }

    private getReadScanQueryForScanIds(scanIds: string[]): string {
        const orConditions = scanIds.map(scanId => {
            return `c.id = '${scanId}'`;
        });

        return `select * from c where ${orConditions.join(' or ')}`;
    }

    private setSystemProperties(pageScanResult: OnDemandPageScanResult): void {
        pageScanResult.itemType = ItemType.onDemandPageScanRunResult;
        pageScanResult.partitionKey = this.getPartitionKey(pageScanResult.id);
    }

    private getPartitionKey(scanId: string): string {
        const node = this.guidUtils.getGuidNode(scanId);

        return this.hashGenerator.getDbHashBucket(OnDemandPageScanRunResultProvider.partitionKeyPreFix, node);
    }
}
