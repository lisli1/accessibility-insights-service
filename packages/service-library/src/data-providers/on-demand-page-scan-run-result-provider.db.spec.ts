// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import 'reflect-metadata';

import { CosmosContainerClient } from 'azure-services';
import { GuidUtils, HashGenerator } from 'common';
import { cloneDeep } from 'lodash';
import { Logger } from 'logger';
import { ItemType, OnDemandPageScanResult } from 'storage-documents';
import { Mock } from 'typemoq';
import { DbMockHelper } from '../test-utilities/db-mock-helpers';
import { OnDemandPageScanRunResultProvider } from './on-demand-page-scan-run-result-provider';

// tslint:disable: no-any

describe('OnDemandPageScanRunResultProvider.Db', () => {
    // tslint:disable-next-line: mocha-no-side-effect-code
    const dbHelper = new DbMockHelper();

    it('no-op', () => {
        // this test exists to have at least 1 test in the test suite to avoid jest failure, when db test run is not supported.
    });

    // tslint:disable-next-line: mocha-no-side-effect-code
    if (dbHelper.isDbTestSupported()) {
        let testSubject: OnDemandPageScanRunResultProvider;

        beforeAll(async () => {
            await dbHelper.init('test-db', 'page-scan-run-results');
        }, 30000);

        beforeEach(() => {
            const loggerMock = Mock.ofType<Logger>();
            const cosmosContainerClient = new CosmosContainerClient(
                dbHelper.cosmosClient,
                dbHelper.dbContainer.dbName,
                dbHelper.dbContainer.collectionName,
                loggerMock.object,
            );

            testSubject = new OnDemandPageScanRunResultProvider(new HashGenerator(), new GuidUtils(), cosmosContainerClient);
        });

        afterEach(async () => {
            await dbHelper.deleteAllDocuments();
        });

        it('stores & retrieve scan results sorted by priority', async () => {
            const partitionKey1Guid1 = '1e9d35cb-ce0c-6590-4455-6b753f5e8e18';
            const partitionKey1Guid2 = '1e9d35cc-a3d0-6780-5a65-6b753f5e8e18';
            const partitionKey2Guid1 = '1e9d35cc-ef1b-65f0-9b5b-b4a0d6006e92';
            const result1: OnDemandPageScanResult = {
                id: partitionKey1Guid1,
                partitionKey: undefined,
                url: 'url1',
                scanResult: { state: 'unknown' },
                reports: [{ reportId: 'rep-id1', href: 'ref1', format: 'sarif' }],
                run: { state: 'failed' },
                priority: 1,
                itemType: ItemType.onDemandPageScanRunResult,
            };
            const result2: OnDemandPageScanResult = {
                id: partitionKey2Guid1,
                partitionKey: undefined,
                url: 'url3',
                scanResult: { state: 'pass' },
                reports: [],
                run: { state: 'completed' },
                priority: 2,
                itemType: ItemType.onDemandPageScanRunResult,
            };

            const result3: OnDemandPageScanResult = {
                id: partitionKey1Guid2,
                partitionKey: undefined,
                url: 'url2',
                scanResult: { state: 'unknown' },
                reports: undefined,
                run: { state: 'running' },
                priority: 2,
                itemType: ItemType.onDemandPageScanRunResult,
            };

            await testSubject.createScanRuns([result1, result2, result3]);

            const itemsInDb = await testSubject.readScanRuns([partitionKey1Guid1, partitionKey2Guid1, partitionKey1Guid2]);

            expect(itemsInDb.length).toBe(3);

            maskSystemProperties(itemsInDb);
            expect(itemsInDb.sort((item1, item2) => (item1.id > item2.id ? 1 : -1))).toMatchSnapshot();
        });

        it('update scan run', async () => {
            const result: OnDemandPageScanResult = {
                id: '1e9d35ce-d95f-62f0-7f39-36c97852967f',
                partitionKey: undefined,
                url: 'url1',
                scanResult: { state: 'unknown' },
                reports: [{ reportId: 'rep-id1', href: 'ref1', format: 'sarif' }],
                run: { state: 'failed' },
                priority: 1,
                itemType: ItemType.onDemandPageScanRunResult,
            };

            const resultUpdate = cloneDeep(result);
            resultUpdate.scanResult = { state: 'pass', issueCount: 3 };

            await testSubject.createScanRuns([result]);
            await testSubject.updateScanRun(result);
            const itemsInDb = await testSubject.readScanRuns([result.id]);

            maskSystemProperties(itemsInDb);
            expect(itemsInDb).toMatchSnapshot();
        });

        function maskSystemProperties(results: OnDemandPageScanResult[]): void {
            results.forEach(result => {
                const systemProperties = Object.keys(result).filter(key => key.startsWith('_'));

                systemProperties.forEach(key => {
                    (result as any)[key] = undefined;
                });
            });
        }
    }
});
