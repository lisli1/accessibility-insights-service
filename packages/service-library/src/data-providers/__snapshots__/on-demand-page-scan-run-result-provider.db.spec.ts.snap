// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnDemandPageScanRunResultProvider.Db stores & retrieve scan results sorted by priority 1`] = `
Array [
  Object {
    "_attachments": undefined,
    "_etag": undefined,
    "_rid": undefined,
    "_self": undefined,
    "_ts": undefined,
    "id": "1e9d35cb-ce0c-6590-4455-6b753f5e8e18",
    "itemType": "pageScanRunResult",
    "partitionKey": "pageScanRunResult-605",
    "priority": 1,
    "reports": Array [
      Object {
        "format": "sarif",
        "href": "ref1",
        "reportId": "rep-id1",
      },
    ],
    "run": Object {
      "state": "failed",
    },
    "scanResult": Object {
      "state": "unknown",
    },
    "url": "url1",
  },
  Object {
    "_attachments": undefined,
    "_etag": undefined,
    "_rid": undefined,
    "_self": undefined,
    "_ts": undefined,
    "id": "1e9d35cc-a3d0-6780-5a65-6b753f5e8e18",
    "itemType": "pageScanRunResult",
    "partitionKey": "pageScanRunResult-605",
    "priority": 2,
    "run": Object {
      "state": "running",
    },
    "scanResult": Object {
      "state": "unknown",
    },
    "url": "url2",
  },
  Object {
    "_attachments": undefined,
    "_etag": undefined,
    "_rid": undefined,
    "_self": undefined,
    "_ts": undefined,
    "id": "1e9d35cc-ef1b-65f0-9b5b-b4a0d6006e92",
    "itemType": "pageScanRunResult",
    "partitionKey": "pageScanRunResult-3",
    "priority": 2,
    "reports": Array [],
    "run": Object {
      "state": "completed",
    },
    "scanResult": Object {
      "state": "pass",
    },
    "url": "url3",
  },
]
`;

exports[`OnDemandPageScanRunResultProvider.Db update scan run 1`] = `
Array [
  Object {
    "_attachments": undefined,
    "_etag": undefined,
    "_rid": undefined,
    "_self": undefined,
    "_ts": undefined,
    "id": "1e9d35ce-d95f-62f0-7f39-36c97852967f",
    "itemType": "pageScanRunResult",
    "partitionKey": "pageScanRunResult-995",
    "priority": 1,
    "reports": Array [
      Object {
        "format": "sarif",
        "href": "ref1",
        "reportId": "rep-id1",
      },
    ],
    "run": Object {
      "state": "failed",
    },
    "scanResult": Object {
      "state": "unknown",
    },
    "url": "url1",
  },
]
`;
